### PR TITLE
SI and OTE WFE updates

### PR DESCRIPTION
**Pull request by mperrin**
PR for code review as suggested by @josePhoenix

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -61,7 +61,7 @@
 builtins._ASTROPY_PACKAGE_NAME_ = PACKAGENAME
 
 # VERSION should be PEP386 compatible (http://www.python.org/dev/peps/pep-0386)
-VERSION = '0.5.0.dev'
+VERSION = '0.6.0.rc1'
```
Comment (josePhoenix) [inline setup.py]: Suggestion: change to `0.6.0rc1` for consistency (e.g. with https://www.python.org/dev/peps/pep-0440/#public-version-identifiers)

File: webbpsf/optics.py [hunk 2]
```diff
@@ -2,17 +2,284 @@
...
+    npix : integer
+        Number of pixels per side of the array
+    outside : float
+        Value to fill the array with outside of the valid segment.
+
+    """
+    from .webbpsf_core import segname
+
+    aper = JWST_primary_aperture(label_segments=True)
+    w = poppy.Wavefront(npix=npix,diam=JWST_primary_aperture.circumscribed_diameter)
```
Comment (josePhoenix) [inline webbpsf/optics.py]: PEP8: need space on the right of a comma

File: webbpsf/optics.py [hunk 3]
```diff
@@ -2,17 +2,284 @@
...
+    segmask = aper.get_transmission(w)
+
+    segname = segname(segnum)
+    cenx, ceny = aper.seg_centers[segname]
+
+    seg_radius = 1.517/2 # nominal point to point diam for A and B segments;
```
Comment (josePhoenix) [inline webbpsf/optics.py]: PEP8: need space on either side of (non-unary) arithmetic operators

File: webbpsf/optics.py [hunk 4]
```diff
@@ -2,17 +2,284 @@
...
-from .webbpsf_core import _log
 
+import logging
+_log = logging.getLogger('webbpsf')
+
+
+
+#######  Classes for modeling aspects of JWST's segmented active primary #####
+
+def segment_zernike_basis(segnum=1, nterms=15, npix=512, outside = np.nan):
```
Comment (josePhoenix) [inline webbpsf/optics.py]: PEP8: no space on either side of `=` in keyword argument definition

File: webbpsf/optics.py [hunk 5]
```diff
@@ -2,17 +2,284 @@
...
+    wg = np.where(segmask==segnum)
+    outzerns = np.full((nterms, npix,npix), outside, dtype=float)
+    outzerns_tmp = poppy.zernike.zernike_basis(nterms=nterms,rho=r[wg], theta=theta[wg], outside=outside)
+    for iz in range(nterms):
+        outzerns[iz][wg] = outzerns_tmp[iz]
+
+    return outzerns
+
+
+class JWST_primary_aperture(poppy.AnalyticOpticalElement):
```
Comment (josePhoenix) [inline webbpsf/optics.py]: For PEP8 compliance, this should be `JWSTPrimaryAperture` (If you ask some of the LSST people, maybe `JwstPrimaryAperture`... but that looks worse to me.)

File: webbpsf/optics.py [hunk 6]
```diff
@@ -2,17 +2,284 @@
...
+
+
+    This class has no wavefront errors, it's just the pupil geometry including
+    the segments (which are not quite perfect hexagons for manufacturing reasons
+    related to trying to tile a curved surface with hexagons while maintaining
+    uniform clearance between adjacent segments) and the secondary mirror support
+    struts, including the bumps on the +V3 strut for the mid boom hinge and mag dampers.
+
+
+    *** CAUTIONARY NOTE ***: At high sampling factors, PSF calculations become a LOT slower.
```
Comment (josePhoenix) [inline webbpsf/optics.py]: I'm not sure if we're following a docstring style, but I favor "NumPy style". I'd change this to

```rest
.. warning:: At high sampling factors, PSF calculations become a LOT slower.
```

(ref: https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt#other-points-to-keep-in-mind)

File: webbpsf/optics.py [hunk 7]
```diff
@@ -2,17 +2,284 @@
...
+
+        segpaths = {}
+        strutpaths=[]
+        for segname, vertices in self.segdata:
+            segpaths[segname] = matplotlib.path.Path(vertices)
+        for strutname, vertices in self.strutdata:
+            strutpaths.append(matplotlib.path.Path(vertices))
+
+        y,x = wave.coordinates()
+        pts = np.asarray([a for a in zip(x.flat, y.flat)])
```
Comment (josePhoenix) [inline webbpsf/optics.py]: I'd use `np.stack` here:

```python
In [37]: %timeit pts = np.asarray([a for a in zip(x.flat, y.flat)])
1 loop, best of 3: 495 ms per loop

In [38]: %timeit pts2 = np.stack((x.flat, y.flat), axis=1)
100 loops, best of 3: 4.8 ms per loop
```

File: webbpsf/optics.py [hunk 8]
```diff
@@ -2,17 +2,284 @@
...
+            self.tel_coords = instrument._tel_coords()
+        else:
+            self.instrument = None
+            self.instr_name = "NIRCam"
+            # figure out default V2V3 coords here
+            self.tel_coords = (0,0) #?
+
+        # determine filename for pupil amplitude array
+        aperture_file = '../jwst_pupil_revW_npix1024.fits.gz'
+        aperture_file =  os.path.abspath(os.path.join(self._datapath,aperture_file))
```
Comment (josePhoenix) [inline webbpsf/optics.py]: This uses an OS-dependent path separator and walks up the tree from the `self._datapath`. I'd change it to:

```python
aperture_file = os.path.abspath(os.path.join(self._WebbPSF_basepath, 'jwst_pupil_revW_npix1024.fits.gz'))
```

File: webbpsf/optics.py [hunk 9]
```diff
@@ -2,17 +2,284 @@
...
+
+        # determine filename for the OPD array
+        #   This should contain a precomputed combination of
+        #   Rev G high spatial frequencies and 
+        #   Rev W mid spatial frequencies
+        # Depends on what the 'level' parameter is.
+
+        if level=='perfect':
+            opd_file = '../OPD_jwst_ote_perfectly_aligned.fits'
```
Comment (josePhoenix) [inline webbpsf/optics.py]: Likewise, I'd use `join()` and `self._WebbPSF_basepath` here.

**Comment by josePhoenix** [comment]
Meta: Hmm. Looks like the commit/diff comments are better for iterating on PRs than these 'reviews'... outdated comments are not hidden automatically and can't be marked resolved as far as I can see.

**Comment by josePhoenix** [comment]
@mperrin if you have no objections, I'm going to go ahead and merge this and the PSF library maker

**Comment by mperrin** [comment]
@josePhoenix thanks for the reminder. sure, please go ahead and merge these PRs. Thanks!

**Pull request merged**